### DISCUSSION TITLE
The elements of a json attribute that are not in the rules function are removed after saving or updating, in the latest v2.6.5.

### DISCUSSION THREAD

**Discussion by ighmouraceneb** [livewire/livewire#3830] [7. Bugs (archived)]
I had a problem since I updated livewire to v2.6.5.
Indeed, I have in a model an attribute of type json : 

```
class View extends Model
{

    use HasFactory;

    protected $fillable = ['name', 'slug', 'description', 'user_id'];

    protected $casts = [
        'config' => 'json'
    ];
}
```
The json object is composed of :

```
{
    "page": {
        "menu": "1",
        "path": "les-recettes",
        "pager": "Full",
        "title": "recettes",
        "fields": "title",
        "isPage": true,
        "withMenu": true,
        "alignment": "horizontal",
        "linkTitle": "les-recettes",
        "htmlListType": "unorder_list",
        "numberColumns": "3",
        "numberPagerPage": "3",
        "numberItemsLimit": "0",
        "displayFormatPage": "grid"
    },
    "show": {
        "type": "1",
        "fields": {
            "title": {
                "title": true,
                "value": "Title"
            }
        },
        "content": "content",
        "rowStyle": "content",
        "rowOptions": "full_content"
    }
}
```

I modify in a component some elements of this object and I define the validation rules in the rules function like this: 

```
<?php

namespace App\Http\Livewire\Back\Structure\Vue;

use App\Models\View;
use Livewire\Component;

class EditParametersDisplayFormatPage extends Component
{

    public $editParametersDisplayFormatPageModal = false;
    public View $view;

    protected $listeners = [
        'editParametersDesplayStyle',
        'formatUpdated'
    ];


    public function rules()
    {
        return [
            'view.config.page.htmlListType' => 'sometimes',
            'view.config.page.alignment' => 'sometimes',
            'view.config.page.numberColumns' => 'required',
            'view.config.page.numberItemsLimit' => 'required'
        ];
    }

}
```
My problem, when I save, I lose all the elements which are not present in the rules function, whereas it was not the case with v2.5.1 of livewire .

With v2.6.5 : 
![image](https://user-images.githubusercontent.com/17543179/134149361-8b014d45-c6f8-46b4-8e51-dae074444889.png)

with v2.5.1  : 
![image](https://user-images.githubusercontent.com/17543179/134148724-7276f805-c578-40ec-99e7-54b3794a384e.png)

do you have an explanation?

Thanks

**Comment by joshhanley** [comment]
@ighmouraceneb for security reasons, any model binding requires a rule to white list attributes. The previous implementation didn't have support for deep attribute binding so in my opinion, that was a bug in the original implementation, which was unintentionally exposing all the json column data to the front end, when it should only have exposed what you had white listed.

But even if only a couple of properties are exposed to the front end, it should still merge the two together, so it could be a bug in the current implementation that is causing it to overwrite your json column data.

I will run an experiment locally tomorrow and see if I can work out what is going on.

Thanks!

**Comment by joshhanley** [comment]
@ighmouraceneb have investigated this today, and it's definitely an issue with merging of dirty data into the models.

The issue is, one of our existing tests ensures that if you modify a particular key in a json attribute with a whole new data set, then only the new data should be kept, and as such does a replace. So I wrote the feature to ensure that existing test passes.

But in your situation it needs to do a recursive merge instead to work properly.

Will have to investigate further on how best to handle it.

**Comment by ighmouraceneb** [comment]
Thanks @joshhanley,
In the meantime, to overcome this problem,
    

-  assign the json column to an array
- make changes to the array
- assign the array back to the json column

```
  public function save()
    {
        
        $this->validate();

        $config = $this->view->config;

        $config['page']['htmlListType'] = $this->htmlListType;
        $config['page']['alignment'] = $this->alignment;
        $config['page']['numberColumns'] = $this->numberColumns;
        $config['page']['numberItemsLimit'] = $this->numberItemsLimit;

        $this->view->config = $config;   

        $this->view->save();
        $this->editParametersDisplayFormatPageModal = false;
        
        $this->emit('parametersStyleUpdated');
    }
```
